File: sdks/php/src/ExecuteRequest.php
<?php

declare(strict_types=1);

namespace Beatbox;

use Beatbox\Internal\Coerce;

/**
 * Body for {@see Client::execute()} and {@see Client::createJob()}.
 *
 * Optional fields are omitted from the wire form unless explicitly set.
 * {@code input} may be any JSON value (including {@code null}), so a
 * private sentinel distinguishes "no input" from "input is null".
 */
final class ExecuteRequest
{
    /** Sentinel meaning "input was not provided". */
    public const UNSET = "\0beatbox:unset\0";

    /**
     * @param mixed $input any JSON value, or {@see self::UNSET} to omit
     */
    public function __construct(
        public Lane $lane,
        public Source $source,
        public ?string $entrypoint = null,
        public mixed $input = self::UNSET,
        public ?string $stdin = null,
        public ?Policy $policy = null,
        public ?string $idempotencyKey = null,
    ) {
    }

    /**
     * Ergonomic constructor for the common wasm-WAT case.
     *
     * @param mixed $input any JSON value, or {@see self::UNSET} to omit
     */
    public static function wasmWat(
        string $text,
        mixed $input = self::UNSET,
        ?string $entrypoint = null,
        ?Policy $policy = null,
    ): self {
        return new self(
            lane: Lane::Wasm,
            source: Source::wasmWat($text),
            entrypoint: $entrypoint,
            input: $input,
            policy: $policy,
        );
    }

    /** @return array<string,mixed> */
    public function toArray(): array
    {
        $out = [
            'lane' => $this->lane->value,
            'source' => $this->source->toArray(),
        ];
        if ($this->entrypoint !== null) {
            $out['entrypoint'] = $this->entrypoint;
        }
        if ($this->input !== self::UNSET) {
            $out['input'] = $this->input;
        }
        if ($this->stdin !== null) {
            $out['stdin'] = $this->stdin;
        }
        if ($this->policy !== null) {
            $out['policy'] = $this->policy->toArray();
        }
        if ($this->idempotencyKey !== null) {
            $out['idempotency_key'] = $this->idempotencyKey;
        }
        return $out;
    }

    /** @param array<string,mixed> $data */
    public static function fromArray(array $data): self
    {
        $lane = Lane::tryFrom(Coerce::stringOr($data['lane'] ?? null, '')) ?? Lane::Wasm;
        $source = Source::fromArray(Coerce::arrayOr($data['source'] ?? null));
        $policy = isset($data['policy']) && is_array($data['policy'])
            ? Policy::fromArray($data['policy'])
            : null;

        return new self(
            lane: $lane,
            source: $source,
            entrypoint: Coerce::stringOrNull($data['entrypoint'] ?? null),
            input: array_key_exists('input', $data) ? $data['input'] : self::UNSET,
            stdin: Coerce::stringOrNull($data['stdin'] ?? null),
            policy: $policy,
            idempotencyKey: Coerce::stringOrNull($data['idempotency_key'] ?? null),
        );
    }
}

File: sdks/java/src/main/java/ai/beatbox/model/ExecuteRequest.java
package ai.beatbox.model;

import com.fasterxml.jackson.annotation.JsonInclude;

/**
 * A request to run a program, used by both {@code execute} and {@code createJob}.
 *
 * <p>Only {@code lane} and {@code source} are required. Use the {@code wasmWat} factories for the
 * common one-liner, or {@link #builder(Lane, Source)} for full control. Null fields are omitted.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ExecuteRequest(
        Lane lane,
        Source source,
        String entrypoint,
        Object input,
        String stdin,
        Policy policy,
        String idempotencyKey) {

    /** Run WAT source on the wasm lane. */
    public static ExecuteRequest wasmWat(String wat) {
        return new ExecuteRequest(Lane.WASM, Source.wasmWat(wat), null, null, null, null, null);
    }

    /** Run WAT source on the wasm lane with the given JSON input. */
    public static ExecuteRequest wasmWat(String wat, Object input) {
        return new ExecuteRequest(Lane.WASM, Source.wasmWat(wat), null, input, null, null, null);
    }

    public static Builder builder(Lane lane, Source source) {
        return new Builder(lane, source);
    }

    /** Fluent builder for requests that set more than the required fields. */
    public static final class Builder {
        private final Lane lane;
        private final Source source;
        private String entrypoint;
        private Object input;
        private String stdin;
        private Policy policy;
        private String idempotencyKey;

        private Builder(Lane lane, Source source) {
            this.lane = lane;
            this.source = source;
        }

        public Builder entrypoint(String entrypoint) {
            this.entrypoint = entrypoint;
            return this;
        }

        public Builder input(Object input) {
            this.input = input;
            return this;
        }

        public Builder stdin(String stdin) {
            this.stdin = stdin;
            return this;
        }

        public Builder policy(Policy policy) {
            this.policy = policy;
            return this;
        }

        public Builder idempotencyKey(String idempotencyKey) {
            this.idempotencyKey = idempotencyKey;
            return this;
        }

        public ExecuteRequest build() {
            return new ExecuteRequest(lane, source, entrypoint, input, stdin, policy, idempotencyKey);
        }
    }
}
